Java/core/src/main/java/com/amazon/randomcutforest/store/PointStore.java

```diff
@@ -290,135 +298,533 @@ public int getDimensions() {
+
+    BitSet inUse() {
+        BitSet result = new BitSet(currentStoreCapacity * dimensions / baseDimension);
+
+        for (int i = 0; i < indexManager.capacity; i++) {
+            if (indexManager.occupied.get(i)) {
+                validateInternalState(refCount[i] > 0, " bits out of sync when count = 0 ");
+                validateInternalState(locationList[i] != PointStore.INFEASIBLE_POINTSTORE_INDEX,
+                        " incorrect location info");
+                result.set(locationList[i] / baseDimension);
+            } else {
+                validateInternalState(refCount[i] == 0, "bits out of sync with count > 0 ");
+                locationList[i] = INFEASIBLE_POINTSTORE_LOCATION;
+            }
+        }
+        return result;
+    }
+
     /**
      * The following function eliminates redundant information that builds up in the
-     * pointstore and shrinks the pointstore
+     * point store and shrinks the point store
      */
 
     public void compact() {
-        checkArgument(!directLocationMap, "incorrect call; should not be used for direct location maps");
-        int DEFAULT_EMPTY = 0; // indicates location not in use
 
         int runningLocation = 0;
         startOfFreeSegment = 0;
-        int stepDimension = (shingleAwareOverlapping) ? baseDimension : dimensions;
-
-        // we first determine which locations are the startpoints of the shingles
-        int[] reverseReferenceList = new int[capacity * dimensions / stepDimension];
-        for (int i = 0; i < capacity; i++) {
-            if (occupied.get(i)) {
-                reverseReferenceList[locationList[i] / stepDimension] = i + 1;
-                // offset by 1, to distinguish DEFAULT_EMPTY
-            }
-        }
 
-        // we make a pass over the data
-        while (runningLocation < capacity * dimensions) {
-            // find the first eligible shoingle to be copied
-            while (runningLocation < capacity * dimensions
-                    && reverseReferenceList[runningLocation / stepDimension] == DEFAULT_EMPTY) {
-                runningLocation += stepDimension;
+        // we first determine which locations are the start points of the shingles
+        // since the shingles extend for a length and can overlap this help define the
+        // region that should be copied
+
+        HashMap<Integer, Integer> movedTo = new HashMap<>();
+
+        // the bit set corresponds to the locations that can be in use over the actual
+        // store array
+        // this is not the same as the number of points that can be stored
+        BitSet inUse = inUse();
+
+        // we make a pass over the store data
+        while (runningLocation < currentStoreCapacity * dimensions) {
+            // find the first eligible shingle to be copied
+            // for rotationEnabled, this should be a multiple of dimensions
+            while (runningLocation < currentStoreCapacity * dimensions && !inUse.get(runningLocation / baseDimension)) {
```

Comment:
Looks like we might be able to use `inUse.nextSetBit()` here.

Reply:
Done